exchange/src/utils/logger/logger.cpp

```diff
@@ -17,35 +19,47 @@ Logger::get_logger()
     return logger;
 }
 
+std::string
+timestamp_in_ms()
+{
+    using namespace std::chrono;
+
+    // Get time from Quill
+    auto now = quill::Clock::now();
+
+    // Get epoch time in milliseconds, and create time point
+    auto epoch_millis = duration_cast<milliseconds>(now.time_since_epoch());
+    auto system_now = time_point<system_clock>{epoch_millis};
+
+    return fmt::format("{:%Y-%m-%d %H:%M:%S}", system_now).substr(0, 23);
+}
+
+template <GlazeMetaSpecialized T>
 void
-Logger::log_event(
-    MESSAGE_TYPE type, const std::string& json_message,
-    const std::optional<std::string>& uid
-)
+Logger::log_event(const T& json_message)
 {
     // If file is not open, throw an error to the error logger
     if (!output_file_.is_open()) [[unlikely]] {
         log_e(events, "Output file {} not open, unable to log event", get_file_name());
         return;
     }
 
-    // Write start of JSON
-    output_file_ << "{ ";
-
-    // Write current GMT time
-    const auto now = std::chrono::system_clock::now();
-    output_file_ << fmt::format("\"time\": \"{:%FT%TZ}\", ", now);
+    WithTimestamp<T> json_message_with_ts = {json_message};
 
-    // Add MessageType and JSON message (and opt UID) to file
-    output_file_ << "\"type\": " << static_cast<int>(type) << ", "; // add type
-    output_file_ << "\"message\": " << json_message;                // add message
+    // TODO: fix type
+    std::string buffer = glz::write_json(json_message_with_ts);
 
-    if (uid.has_value()) {
-        output_file_ << ", \"uid\": " << uid.value(); // add uid if exists
-    }
+    // Log to the main log too
+    log_i(events, "Logging event {}", buffer);
 
-    output_file_ << " }\n"; // close the brace and end the line
+    // Write start of JSON
+    output_file_ << buffer << "\n";
 }
 
+// Explicit instantiations
```

Comment:
Why explicit?

Reply:
steve did this, I have no idea

Reply:
Because we're separating header and implementation of a template function

Reply:
Again

W H Y

Reply:
Because we know the types that the template will be instantiated for so why not?

Reply:
Fine